### PR TITLE
Add the PVRL method (rebased)

### PR DESCRIPTION
**Pull request by arthurPignet**
Following the (PR 180)[https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/180]

Need to be merged after the add imdb-dataset-V2

### CODE REVIEW

**Review by Thomas-Galtier — APPROVED** [review]

File: mplc/contributivity.py [hunk 1]
```diff
@@ -938,6 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
...
+                        1.0 - partner_values[i]) - prodp / (1.0 - prodp) / (1.0 - partner_values[i])
+                new_w[i] = w[i] + learning_rate * G * dp_dw[i] * grad
+            w = new_w
+            # Update values before the next round
+            previous_partner_values = partner_values
+            partner_values = np.exp(w) / (1.0 + np.exp(w))
+            previous_loss = loss
+
+        self.name = "PVRL"
+        print(f"PVRL: the final accuracy is  {mpl.test_score}")
```
Comment (natct10) [inline mplc/contributivity.py:1016]: It would be nice to log this instead of printing it

File: mplc/contributivity.py [hunk 2]
```diff
@@ -938,6 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
...
+            use_saved_weights=False,
+        )
+        mpl.compute_test_score()
+        previous_loss = mpl.loss_collective_models[-1]
+        t = 0
+        while (t < the_scenario.epoch_count * the_scenario.partners_count or np.sum(
+                np.abs(partner_values - previous_partner_values)) / the_scenario.partners_count > epsilon):
+            t += 1
+            print("t:", t)
+            print("partner_values:", partner_values)
```
Comment (natct10) [inline mplc/contributivity.py:972]: log it?

File: mplc/multi_partner_learning.py [hunk 3]
```diff
@@ -554,21 +553,10 @@ def init_with_agg_models(self):
 
         self.prepare_aggregation_weights()
         partners_model_list = []
-        for _ in self.partners_list:
+        for partner in self.partners_list:
```
Comment (natct10) [inline mplc/multi_partner_learning.py:556]: +1

File: config_quick_debug.yml [hunk 4]
```diff
@@ -10,14 +10,12 @@ scenario_params_list:
...
      - ['basic', 'random']
-     - ['basic', 'stratified']
-     - ['advanced', [[4, 'shared'], [6, 'shared'], [4, 'specific']]]
    multi_partner_learning_approach:
     - 'fedavg'
    aggregation_weighting:
     - 'uniform'
    methods:
-    - ["Shapley values", "Independent scores"]
+    - ["TMCS", "PVRL"]
```
Comment (natct10) [inline config_quick_debug.yml:18]: Is this file supposed to evolve?

**Review by natct10 — COMMENTED** [review]
Looks cool, thank you @arthurPignet!

File: config_quick_debug.yml (hunk 4 above)
Comment (Thomas-Galtier) [reply config_quick_debug.yml:18]: We said thuesday that  the config_debug can change now, as we will create a config_travis. I may have anticipated this change here. I will put config_debug back as it was.

File: mplc/multi_partner_learning.py [hunk 5]
```diff
@@ -208,14 +209,14 @@ def compute_test_score(self):
         self.test_score = model_evaluation_test_data[1]  # 0 is for the loss
         self.nb_epochs_done = self.epoch_index + 1
 
-        # Plot training history
-        # TODO: move the data saving and plotting in dedicated functions
+        # Plot training history # TODO: move the data saving and plotting in dedicated functions
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:212]: I think this TODO is done, we can remove it

File: mplc/contributivity.py [hunk 6]
```diff
@@ -938,6 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
             end = timer()
             self.computation_time_sec = end - start
 
+    # %% compute Partner value by reinforocment learning
+
+    def PVRL(self, the_scenario, learning_rate):
```
Comment (RomainGoussault) [inline mplc/contributivity.py:945]: Can we provide the link to the PVRL paper ?

File: mplc/multi_partner_learning.py [hunk 7]
```diff
@@ -109,7 +109,8 @@ def compute_test_score_for_single_partner(self, partner):
 
         # Save model score on test data
         self.test_score = model_evaluation_test_data[1]  # 0 is for the loss
-        self.nb_epochs_done = (es.stopped_epoch + 1) if es.stopped_epoch != 0 else self.epoch_count
+        self.loss_collective_models.append(model_evaluation_test_data[0])  # store the loss for PVRL
+        self.nb_epochs_done = (es.stopped_epoch + 1) if self.is_early_stopping else self.epoch_count
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:113]: Looking at Eric's comment here --> https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/141#discussion_r428158162
I think this won't work.

File: mplc/contributivity.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 
 import bisect
 import datetime
+import os
```
Comment (RomainGoussault) [inline mplc/contributivity.py:10]: Import pathlib ;)

**Review by RomainGoussault — CHANGES_REQUESTED** [review]
Nice work !
Can we provide a test for this contributivity method ? An end to end test might be too long to run so maybe we could the PVRL method only ?

File: mplc/contributivity.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Thomas-Galtier) [reply mplc/contributivity.py:945]: There is no PVRL paper yet but only a paper on DVRL

File: mplc/multi_partner_learning.py (hunk 7 above)
Comment (Thomas-Galtier) [reply mplc/multi_partner_learning.py:113]: Actually this line was not working when the earlystooping was deactivated because the `es` object was no more instantiated in that case. This new line also to fix this.

File: mplc/contributivity.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RomainGoussault) [reply mplc/contributivity.py:945]: Then maybe add one or two lines describing the method. Also we should update the doc here: https://github.com/SubstraFoundation/distributed-learning-contributivity/blob/master/mplc/docs/documentation.md#configuration-of-contributivity-measurement-methods-to-be-tested

File: mplc/multi_partner_learning.py (hunk 7 above)
Comment (RomainGoussault) [reply mplc/multi_partner_learning.py:113]: Ok I understand.
However I think Eric's [comment](https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/141#discussion_r428158162) still holds: 

> Via the callback the EarlyStopping provides stopped_epoch. But the way it is implemented, if early stopping wasn't trigger, this returns 0. 


So if `is_early_stopping = true` but the early stopping was never triggered, `stopped_epoch = 0`

File: mplc/contributivity.py [hunk 9]
```diff
@@ -938,6 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
             end = timer()
             self.computation_time_sec = end - start
 
+    # %% compute Partner value by reinforocment learning
```
Comment (natct10) [inline mplc/contributivity.py:943]:
```suggestion
    # %% compute Partner value by reinforcement learning
```

Comment (natct10) [reply mplc/contributivity.py:943]: Is there a meaning for `%%`?

Comment (arthurPignet) [reply mplc/contributivity.py]: I don't know 
Maybe in some IDE ? All the comments about the contributivity measures follow this format

File: mplc/docs/documentation.md [hunk 10]
```diff
@@ -403,7 +338,7 @@ dataset = dataset.Dataset(
 ```
 
 ### Model generator
-This function provides the model which will be trained by the scenario object. Currently the library handles compiled Keras' model (see MNIST, ESC50, IMDB and CIFAR10 datasets), and Scikit-Learn Linear Regression (see the TITANIC dataset).  
+This function provides the model use, which will be trained by the scenario object.
```
Comment (bowni) [inline mplc/docs/documentation.md:341]: This is weird, it looks like `documentation.md` goes back in time versus its current state in `master`?

File: mplc/contributivity.py [hunk 11]
```diff
@@ -938,103 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
             end = timer()
             self.computation_time_sec = end - start
 
-    def federated_SBS_linear(self, the_scenario):
-        start = timer()
-        logger.info(
-            "# Launching computation of perf. scores of linear "
-            "performance increase compared to previous collective model")
-
-        relative_perf_matrix = self.compute_relative_perf_matrix(the_scenario)
```
Comment (bowni) [inline mplc/contributivity.py:947]: Same in this file, it looks like it wasn't well rebased, and it erases quite a lot of recently merged material.

File: mplc/contributivity.py (hunk 1 above)
Comment (bowni) [reply mplc/contributivity.py:1016]: I concur with @natct10 remark 😄 Any reason @arthurPignet @Thomas-Galtier why it was dismissed?

**Review by bowni — CHANGES_REQUESTED** [review]

File: mplc/docs/documentation.md (hunk 10 above)
Comment (natct10) [reply mplc/docs/documentation.md:341]: I think this PR will need to be rebased as you pointed before @bowni

File: mplc/multi_partner_learning.py [hunk 12]
```diff
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (natct10) [inline mplc/multi_partner_learning.py:1]: Why remove the encoding tag? I'm curious

File: mplc/multi_partner_learning.py [hunk 13]
```diff
@@ -102,14 +101,19 @@ def compute_test_score_for_single_partner(self, partner):
...
         # Evaluate trained model
         model_evaluation_test_data = self.evaluate_model(model, self.test_data)
         logger.info(f"   Model evaluation on test data: "
                     f"{list(zip(model.metrics_names, ['%.3f' % elem for elem in model_evaluation_test_data]))}")
 
         # Save model score on test data
         self.test_score = model_evaluation_test_data[1]  # 0 is for the loss
-        self.nb_epochs_done = (es.stopped_epoch + 1) if es.stopped_epoch != 0 else self.epoch_count
+        self.nb_epochs_done = (es.stopped_epoch + 1) if (self.is_early_stopping
+                                                         and es.stopped_epoch != 0) else self.epoch_count
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:116]: cf my previous comment:
https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/241#discussion_r499772512

**Review by RomainGoussault — CHANGES_REQUESTED** [review]

File: mplc/contributivity.py [hunk 14]
```diff
@@ -938,103 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
...
-        # Return contributivity scores
-        self.name = "Federated step by step constant scores"
-        self.contributivity_scores = contributivity_scores
+        mpl.compute_test_score()
+        previous_loss = mpl.loss_collective_models[-1]
+        t = 0
+        while (t < the_scenario.epoch_count * the_scenario.partners_count or np.sum(
+                np.abs(partner_values - previous_partner_values)) / the_scenario.partners_count > epsilon):
+            t += 1
+            logger.info("t:", t)
```
[the remark this answers is not in the document]
Comment (RomainGoussault) [reply mplc/contributivity.py:971]: Actually this only logs the first "t".
We should do something like:             `logger.info(f"t: {t}")`

File: mplc/contributivity.py [hunk 15]
```diff
@@ -938,103 +940,88 @@ def without_replacment_SMC(self, the_scenario, sv_accuracy=0.01, alpha=0.95):
...
+            logger.info("partner_values:", partner_values)
```
Comment (RomainGoussault) [inline mplc/contributivity.py:972]: same here, it will only log the first string

File: mplc/multi_partner_learning.py (hunk 13 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py]: I think I missed something
In the case of there is an early stopping, and it was triggered (ie es.stopped_epoch > 0), the number of epoch done is equal to es.stopped_epoch right ?
At the opposite, if there is no es, or if it was not triggered, the number of epoch done should be epoch_count ?

File: mplc/contributivity.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (arthurPignet) [reply mplc/contributivity.py]: In fact this t represents the number of epoch done. I renamed it PVRL_epoch. Due to the implementation of PVRL, most of the log will make no sense. (as a new mpl object is recreated at each epoch). It will be solved with the refacto of mpl, but I need changes which are done in the refacto to modify that.

**Review by bowni — APPROVED** [review]

File: mplc/contributivity.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bowni) [reply mplc/contributivity.py:944]: There are a few words about it in the `documentation.md`

File: mplc/docs/documentation.md [hunk 16]
```diff
@@ -350,13 +350,13 @@ The methods are detailed below:
   - `["SMCS"]` **Stratified Monte Carlo Shapley with replacement**
   - `["WR_SMC"]` **Stratified Monte Carlo Shapley without replacement**
 
-- [In progress]  **Partner Valuation by Reinforcement Learning** :
+- **Partner Valuation by Reinforcement Learning** :
```
Comment (natct10) [inline mplc/docs/documentation.md:353]:
```suggestion
- **Partner Valuation by Reinforcement Learning**:
```

**Review by natct10 — APPROVED** [review]
Just one suggestion and approved!

**Review by bowni — APPROVED** [review]

**Review by RomainGoussault — APPROVED** [review]
🎉 !!
